### PR TITLE
Fixed #312: implemented RsPrettyJSON

### LINKED ISSUE #312: RsPrettyJSON
Let's create class `RsPrettyJSON` similar to `RsPrettyXML` (see #296), to format JSON nicely

### PR DESCRIPTION
**Pull request by ekondrashev**
Implemented RsPrettyJSON requests at #312 .
Tests are also available.

Implementation was inspired by solution for #296, see PR #308.

### CODE REVIEW

**Comment by ekondrashev** [comment]
@davvd please see this PR fixing #312

**Comment by davvd** [comment]
@ekondrashev Thanks for your pull request, let me find someone who can review it

**Comment by davvd** [comment]
@pinaf review this one,please

File: src/main/java/org/takes/rs/RsPrettyJSON.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+    @Override
+    public InputStream body() throws IOException {
+        return this.make().body();
+    }
+
+    /**
+     * Make a response.
+     * @return Response just made
+     * @throws java.io.IOException If fails
```
Comment (pinaf) [inline src/main/java/org/takes/rs/RsPrettyJSON.java]: @ekondrashev `@throws IOException If fails` is fine

File: src/main/java/org/takes/rs/RsPrettyJSON.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            }
+        }
+        return this.transformed.get(0);
+    }
+
+    /**
+     * Format body with proper indents.
+     * @param body Response body
+     * @return New properly formatted body
+     * @throws java.io.IOException If fails
```
Comment (pinaf) [inline src/main/java/org/takes/rs/RsPrettyJSON.java]: @ekondrashev same here

File: src/main/java/org/takes/rs/RsPrettyJSON.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+     */
+    private static byte[] transform(final InputStream body) throws IOException {
+        final ByteArrayOutputStream result = new ByteArrayOutputStream();
+        final JsonReader reader = Json.createReader(body);
+        final Map<String, ?> props = Collections.singletonMap(
+            JsonGenerator.PRETTY_PRINTING,
+            true
+        );
+        final JsonWriterFactory factory = Json.createWriterFactory(props);
```
Comment (pinaf) [inline src/main/java/org/takes/rs/RsPrettyJSON.java]: @ekondrashev let's inline this variable

File: src/main/java/org/takes/rs/RsPrettyJSON.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        final JsonWriter writer = factory.createWriter(result);
+        try {
```
Comment (pinaf) [inline src/main/java/org/takes/rs/RsPrettyJSON.java]: @ekondrashev let's use the auto-closing try block for the reader and writer variables.

File: src/test/java/org/takes/rs/RsPrettyJSONTest.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import nl.jqno.equalsverifier.EqualsVerifier;
+import nl.jqno.equalsverifier.Warning;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
+import org.junit.Test;
+
+/**
+ * Test case for {@link org.takes.rs.RsPrettyJSON}.
```
Comment (pinaf) [inline src/test/java/org/takes/rs/RsPrettyJSONTest.java]: @ekondrashev `Test case for {@link RsPrettyJSON}.`

File: src/test/java/org/takes/rs/RsPrettyJSONTest.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+ * @author Eugene Kondrashev (eugene.kondrashev@gmail.com)
+ * @version $Id$
+ * @since 1.0
+ */
+public final class RsPrettyJSONTest {
+
+    /**
+     * RsPrettyJSON can format response with JSON body.
+     * @throws java.io.IOException If some problem inside
```
Comment (pinaf) [inline src/test/java/org/takes/rs/RsPrettyJSONTest.java]: @ekondrashev `IOException` is enough

File: src/test/java/org/takes/rs/RsPrettyJSONTest.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+            ).printBody(),
+            Matchers.is(
+                "\n{\n    \"widget\":{\n        \"debug\":\"on\"\n    }\n}"
+            )
+        );
+    }
+
+    /**
+     * RsPrettyJSON can format response with non JSON body.
+     * @throws java.io.IOException If some problem inside
```
Comment (pinaf) [inline src/test/java/org/takes/rs/RsPrettyJSONTest.java]: @ekondrashev same here

File: src/test/java/org/takes/rs/RsPrettyJSONTest.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+                )
+            ).printBody(),
+            Matchers.is(
+                "\n{\n    \"widget\":{\n        \"debug\":\"on\"\n    }\n}"
+            )
+        );
+    }
+
+    /**
+     * RsPrettyJSON can format response with non JSON body.
```
Comment (pinaf) [inline src/test/java/org/takes/rs/RsPrettyJSONTest.java]: @ekondrashev `RsPrettyJSON can reject a non-JSON body`

File: src/test/java/org/takes/rs/RsPrettyJSONTest.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+     * @throws java.io.IOException If some problem inside
+     */
+    @Test(expected = IOException.class)
+    public void formatsNonJsonBody() throws IOException {
```
Comment (pinaf) [inline src/test/java/org/takes/rs/RsPrettyJSONTest.java]: @ekondrashev `rejectsNonJsonBody`

File: src/test/java/org/takes/rs/RsPrettyJSONTest.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        new RsPrint(new RsPrettyJSON(new RsWithBody("foo"))).printBody();
+    }
+
+    /**
+     * RsPrettyJSON can report correct content length.
+     * @throws java.io.IOException If some problem inside
```
Comment (pinaf) [inline src/test/java/org/takes/rs/RsPrettyJSONTest.java]: @ekondrashev same here

**Comment by pinaf** [comment]
@ekondrashev 10 comments above

**Comment by pinaf** [comment]
@ekondrashev let's try to clean up the JSON transformation code a bit by using try-with-resources for auto-closing or reader and writer.

**Comment by pinaf** [comment]
@ekondrashev btw, nice job on the tests :)

File: src/main/java/org/takes/rs/RsPrettyJSON.java (hunk 4 above)
Comment (ekondrashev) [inline src/main/java/org/takes/rs/RsPrettyJSON.java]: If you are saying about `try-with-resource` then it was introduced with java7, but from the Takes readme:
```
Takes is a true object-oriented and immutable Java6 web development framework..
```
Am I missing something?

**Comment by ekondrashev** [comment]
> @ekondrashev btw, nice job on the tests :)
it was @ikhvostenkov nice job initially, I just used his work on `RsPrettyXML`, so @ikhvostenkov comment is for you, I can just +1 :)

**Comment by ekondrashev** [comment]
@pinaf done, with your comments, plz take a look.

File: src/main/java/org/takes/rs/RsPrettyJSON.java (hunk 4 above)
Comment (pinaf) [inline src/main/java/org/takes/rs/RsPrettyJSON.java]: @ekondrashev you are right.

File: src/main/java/org/takes/rs/RsPrettyJSON.java [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+    /**
+     * Format body with proper indents.
+     * @param body Response body
+     * @return New properly formatted body
+     * @throws IOException If fails
+     */
+    private static byte[] transform(final InputStream body) throws IOException {
+        final ByteArrayOutputStream res = new ByteArrayOutputStream();
+        final JsonReader rdr = Json.createReader(body);
```
Comment (pinaf) [inline src/main/java/org/takes/rs/RsPrettyJSON.java]: @ekondrashev let's move this into the first try block.

File: src/main/java/org/takes/rs/RsPrettyJSON.java [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+        final Map<String, ?> prp = Collections.singletonMap(
+            JsonGenerator.PRETTY_PRINTING,
+            true
+        );
+        final JsonWriter wrt = Json.createWriterFactory(prp).createWriter(res);
```
Comment (pinaf) [inline src/main/java/org/takes/rs/RsPrettyJSON.java]: @ekondrashev let's move this into the second try block.

**Comment by pinaf** [comment]
@ekondrashev thanks. 2 more comments above

**Comment by pinaf** [comment]
@ekondrashev also, please make sure to address all of your comments to someone. please go back and check each comment you made in this PR and address it to someone otherwise QA will complain.

File: src/main/java/org/takes/rs/RsPrettyJSON.java (hunk 11 above)
Comment (ekondrashev) [inline src/main/java/org/takes/rs/RsPrettyJSON.java]: @pinaf could you please elaborate on reasons? `Json.createReader` does throw no exceptions.

File: src/main/java/org/takes/rs/RsPrettyJSON.java (hunk 12 above)
Comment (ekondrashev) [inline src/main/java/org/takes/rs/RsPrettyJSON.java]: @pinaf same here, `createWriter` call throws no exceptions. Any reason to move it to try block?

**Comment by pinaf** [comment]
@ekondrashev If an exception is thrown on line 117
```java
final JsonObject obj = rdr.readObject();
```
the writer will never be closed.

**Comment by pinaf** [comment]
@ekondrashev also, it makes the code more readable, don't you think?

**Comment by ekondrashev** [comment]
@pinaf you have a point with regard to the writer, but that means it should be moved to the first try block not the second, correct?. Committed the change. Please take a look if anything else should be changed.

**Comment by pinaf** [comment]
@rultor merge

**Comment by pinaf** [comment]
@ekondrashev thanks for your patience :)

**Comment by pinaf** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rultor merge pls

**Comment by davvd** [comment]
@ypshenychka review this ticket please, for compliance with our [quality rules](http://www.teamed.io/qa.html)

**Comment by davvd** [comment]
@rultor please deploy

**Comment by ypshenychka** [comment]
@ekondrashev 
>If you are saying about try-with-resource then it was introduced with java7, but from the Takes readme:
Takes is a true object-oriented and immutable Java6 web development framework..
Am I missing something?

According to our [QA Rules](http://www.teamed.io/qa.html):
>Messages in a ticket always start with a name of a user they are addressed to.

Please  indicate a person you're referring to in the beginning of your comment.

**Comment by ekondrashev** [comment]
@ypshenychka done, thx!

**Comment by ypshenychka** [comment]
@ekondrashev great, thank you

**Comment by ypshenychka** [comment]
@davvd Quality is acceptable.

**Comment by davvd** [comment]
> @davvd Quality is acceptable.

@ypshenychka thanks for the QA review, we'll work better next time

**Comment by davvd** [comment]
@pinaf I added 10 mins to @ypshenychka (for QA review) in transaction 59087925... thanks, I just added **25 mins** to your account, payment `AP-1G74488364945652A`, 133 hours and 42 mins spent... review comments (c=10) added as a bonus... +25 added to your rating, at the moment it is: [+6976](http://www.netbout.com/b/35128?open=rating)

**Pull request merged**